[extension/filestorage] replace path-unsafe characters in component names

andrzej-stencel opened the pull request:
Fixes #3148

This change was initially created to only fix #20731 by replacing the slash `/` characters in component names with a tilde `~`. After that, we decided that it is a breaking change, and so it's better to fix other characters as well in a single breaking change.

andrzej-stencel committed:
[extension/filestorage] replace path-unsafe characters in component names

Fixes #3148

atoulme commented:
Will this break existing checkpoint file paths, causing existing receivers to lose their history?

astencel-sumo commented:
> Will this break existing checkpoint file paths, causing existing receivers to lose their history?

You're right, this will change the directory used for storage for users already using components with a slash in the name, e.g. for the following configuration:

```yaml
extensions:
  file_storage:
    directory: /tmp

receivers:
  filelog/logs/json:
    storage: file_storage
```

As long as the user has made sure the path `tmp/receiver_filelog_logs/json` exists before running the collector (e.g. with `mkdir -p /tmp/receiver_filelog_logs/json`), this change will change the storage path from `tmp/receiver_filelog_logs/json` to `tmp/receiver_filelog_logs~json`.

I wonder how plausible this scenario is. :thinking: 

Perhaps we should treat this "fix" as a "breaking change" and introduce it via a [feature gate](https://github.com/open-telemetry/opentelemetry-collector/tree/v0.75.0/featuregate)?

djaglowski commented:
Code looks good to me but I do think we need to treat this as a breaking change.

astencel-sumo commented:
> Code looks good to me but I do think we need to treat this as a breaking change.

Fair enough, I'll add a feature gate for this.

astencel-sumo commented:
I'm still going to work on it. I think that if we want to treat this as a breaking change, it would be good to fix not only the slash issue, but also other potential problematic characters in one breaking change.

One example is percent sign `%`, which is problematic on Windows platform, but not on other platforms. To keep things consistent, I think we should replace the `%` in paths on all platforms, not only on Windows? :thinking:

astencel-sumo commented:
Let's keep it open :slightly_smiling_face: I'll be back!

andrzej-stencel committed:
feat: add feature gate

astencel-sumo commented:
I have added a feature gate and rewritten the logic to replace all characters other than a safe set. Please take a look @djaglowski.

astencel-sumo commented on extension/storage/filestorage/README.md:
```diff
@@ -93,3 +93,32 @@ processors:
+The schedule for this feature gate is:
+
+- Introduced in v0.81.0 (July 2023) as `alpha` - disabled by default.
+- Moved to `beta` in October 2023 - enabled by default.
```
I described the schedule for the feature gate in terms of dates, not release numbers, as this is pretty far in the future and I believe it's easier for everyone to focus on time frames as opposed to arbitrary release numbers.

The schedule I selected is slow: three months before moving the gate to `beta` and another three months between marking it `stable`. This is because the component itself is marked as `beta`, so users might expect changes to not happen rapidly. Also I believe there's no rush - the feature gate is available, so users can switch it on if the original behavior bothers them.

Let me know what you think!

andrzej-stencel committed:
test: fix unit tests after introducing feature gate

andrzej-stencel committed:
make gotidy

djaglowski reviewed (approved):
Thanks @astencel-sumo!

tigrannajaryan commented on extension/storage/filestorage/README.md:
```diff
@@ -93,3 +93,32 @@ processors:
+For example, for component `filelog/logs/json`, the data is stored:
+
+- in path `receiver_filelog_logs/json` with the feature flag disabled (note that this is file named `json` inside directory named `receiver_filelog_logs`).
+- in file `receiver_filelog_logs~json` with the feature flag enabled.
```
Wouldn't it be better to encode unsafe characters in terms of safe characters instead of replacing? Replacing creates the possibility of name clashes, e.g. components `filelog/logs/json` and `filelog/logs~json` both convert to `receiver_filelog_logs~json`.

astencel-sumo replied:
That's true, the issue of clashes crossed my mind. Especially with the replacement being quite "eager", replacing many different characters. I'll change the implementation to encode unsafe characters instead of replacing with `~`.

newly12 replied:
@astencel-sumo may I ask if you get any chance to work on this?

astencel-sumo replied:
@newly12 the vacation period is over and I'm going to get back to it.

MovieStoreGuy commented:
Can I ask, why not just error when there is a `component.Name()` that is invalid?

djaglowski commented:
> Can I ask, why not just error when there is a `component.Name()` that is invalid?

We're generating a filename based on several fields, including a component ID, so we shouldn't error on anything characters that could be part of a valid component ID.

andrzej-stencel committed:
Merge branch 'main' into replace-slashes-in-component-names

andrzej-stencel committed:
feat: encode unsafe characters instead of replacing

This prevents clashes between similar unsafe names, e.g.:
- `filelog/logs/json`
- `filelog/logs~json`

astencel-sumo commented:
@tigrannajaryan I have reworked the implementation to encode unsafe characters instead of replacing. Please take a look.

andrzej-stencel committed:
Merge branch 'main' into replace-slashes-in-component-names

andrzej-stencel committed:
Merge branch 'main' into replace-slashes-in-component-names
